learn/chapter1/4-shaders.md

```diff
@@ -0,0 +1,272 @@
+
+# GLSL
+
+ Shaders are written in the C-like language GLSL. GLSL is tailored for use with graphics and contains useful features specifically targeted at vector and matrix manipulation.
+
+Shaders always begin with a version declaration, followed by a list of input and output variables, uniforms and its main function. Each shader's entry point is at its `main` function where we process any input variables and output the results in its output variables. Don't worry if you don't know what uniforms are, we'll get to those shortly.
+
+A shader typically has the following structure: 
+
+```glsl
+#version version_number
+in type in_variable_name;
+in type in_variable_name;
+
+out type out_variable_name;
+  
+uniform type uniform_name;
+  
+void main()
+{
+  // process input(s) and do some weird graphics stuff
+  ...
+  // output processed stuff to output variable
+  out_variable_name = weird_stuff_we_processed;
+}
+```
+When we're talking specifically about the vertex shader each input variable is also known as a `vertex attribute`. There is a maximum number of vertex attributes we're allowed to declare limited by the hardware. OpenGL guarantees there are always at least 16 4-component vertex attributes available, but some hardware may allow for more which you can retrieve by querying `GL_MAX_VERTEX_ATTRIBS:`
+
+```cs 
+int nrAttributes = 0;
+GL.GetInteger(GetPName.MaxVertexAttribs, out nrAttributes);
+Console.WriteLine("Maximum number of vertex attributes supported: " + nrAttributes);
+```
+This often returns the minimum of 16 which should be more than enough for most purposes.
+
+## Types
+
+GLSL has, like any other programming language, data types for specifying what kind of variable we want to work with. GLSL has most of the default basic types we know from languages like C: int, float, double, uint and bool. GLSL also features two container types that we'll be using a lot, namely vectors and matrices. We'll discuss matrices in a later chapter. 
+
+### Vectors
+
+ A vector in GLSL is a 1,2,3 or 4 component container for any of the basic types just mentioned. They can take the following form (n represents the number of components):
+
+ - vecn: the default vector of n floats.
+ - bvecn: a vector of n booleans.
+ - ivecn: a vector of n integers.
+ - uvecn: a vector of n unsigned integers.
+ - dvecn: a vector of n double components.
+
+Most of the time we will be using the basic vecn since floats are sufficient for most of our purposes.
+
+Components of a vector can be accessed via vec.x where x is the first component of the vector. You can use .x, .y, .z and .w to access their first, second, third and fourth component respectively. GLSL also allows you to use rgba for colors or stpq for texture coordinates, accessing the same components.
+
+The vector datatype allows for some interesting and flexible component selection called swizzling. Swizzling allows us to use syntax like this:
+
+``` glsl
```

Comment:
Will the markdown work with a space here?

Reply:
Yeah I looked at it in visual studio code and the rendering in there seemed fine

Reply:
:+1: